File: apps/api/src/middleware/teacherScope.ts
import type { NextFunction, Response } from "express";
import { pool } from "../db/pool.js";
import { normalizeUserRole } from "../db/userSql.js";
import type { AuthenticatedTenantRequest } from "./tenantAuth.js";

export interface TeacherScopedRequest extends AuthenticatedTenantRequest {
  /** null means all classes in the school are allowed (admin / head teacher). */
  allowedClassIds?: string[] | null;
}

export async function teacherScope(
  req: TeacherScopedRequest,
  res: Response,
  next: NextFunction,
) {
  const schoolId = req.schoolId;
  const role = req.tenantUser?.role;

  if (!schoolId || !role) {
    return res.status(401).json({ error: "Not authenticated" });
  }

  const normalizedRole = normalizeUserRole(role);

  if (normalizedRole === "admin" || normalizedRole === "head_teacher") {
    req.allowedClassIds = null;
    return next();
  }

  if (normalizedRole !== "teacher") {
    return res.status(403).json({ error: "Forbidden" });
  }

  const result = await pool.query<{ class_id: string }>(
    `SELECT DISTINCT class_id
     FROM teacher_class_assignments
     WHERE school_id = $1 AND teacher_id = $2`,
    [schoolId, req.tenantUser!.sub],
  );

  req.allowedClassIds = result.rows.map((row) => row.class_id);
  next();
}

export function assertClassAccess(req: TeacherScopedRequest, classId: string) {
  if (req.allowedClassIds === null) {
    return true;
  }
  return req.allowedClassIds?.includes(classId) ?? false;
}

File: apps/api/app/middleware/teacher_scope.py
import uuid
from typing import Any

import asyncpg
from fastapi import Depends, HTTPException, status

from app.db.pool import get_db
from app.middleware.tenant import get_tenant_and_user


async def get_allowed_class_ids(
    ctx: tuple[uuid.UUID, dict[str, Any]] = Depends(get_tenant_and_user),
    conn: asyncpg.Connection = Depends(get_db),
) -> list[uuid.UUID] | None:
    school_id, user = ctx
    role = user["role"]
    if role in ("admin", "head_teacher"):
        return None
    if role != "teacher":
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail={"error": "Forbidden"})

    rows = await conn.fetch(
        """
        SELECT DISTINCT class_id
        FROM teacher_class_assignments
        WHERE school_id = $1 AND teacher_id = $2
        """,
        school_id,
        user["user_db_id"],
    )
    return [row["class_id"] for row in rows]


def assert_class_access(allowed_class_ids: list[uuid.UUID] | None, class_id: uuid.UUID) -> bool:
    if allowed_class_ids is None:
        return True
    return class_id in allowed_class_ids
